File: src/main/java/pl/mjedynak/idea/plugins/builder/finder/BuilderFinder.java
package pl.mjedynak.idea.plugins.builder.finder;

import com.intellij.psi.PsiClass;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.VisibleForTesting;

public class BuilderFinder {

    @VisibleForTesting
    static final String SEARCH_PATTERN = "Builder";

    public static final String EMPTY_STRING = "";

    private BuilderFinder() {
        throw new UnsupportedOperationException("Utility class");
    }

    @Nullable
    public static PsiClass findBuilderForClass(@NotNull PsiClass psiClass) {
        PsiClass innerBuilderClass = tryFindInnerBuilder(psiClass);
        if (innerBuilderClass != null) {
            return innerBuilderClass;
        } else {
            String className = psiClass.getName();
            if (className == null) return null;
            return findClass(psiClass, className + SEARCH_PATTERN);
        }
    }

    @Nullable
    private static PsiClass tryFindInnerBuilder(@NotNull PsiClass psiClass) {
        PsiClass innerBuilderClass = null;
        PsiClass[] allInnerClasses = psiClass.getAllInnerClasses();
        for (PsiClass innerClass : allInnerClasses) {
            String innerClassName = innerClass.getName();
            if (innerClassName == null) continue;
            if (innerClassName.contains(SEARCH_PATTERN)) {
                innerBuilderClass = innerClass;
                break;
            }
        }
        return innerBuilderClass;
    }

    @Nullable
    public static PsiClass findClassForBuilder(@NotNull PsiClass psiClass) {
        String className = psiClass.getName();
        if (className == null) return null;
        String searchName = className.replaceFirst(SEARCH_PATTERN, EMPTY_STRING);
        return findClass(psiClass, searchName);
    }

    @Nullable
    private static PsiClass findClass(@NotNull PsiClass psiClass, @NotNull String searchName) {
        PsiClass result = null;
        if (typeIsCorrect(psiClass)) {
            result = ClassFinder.findClass(searchName, psiClass.getProject());
        }
        return result;
    }

    private static boolean typeIsCorrect(@NotNull PsiClass psiClass) {
        return !psiClass.isAnnotationType() && !psiClass.isEnum() && !psiClass.isInterface();
    }
}

File: src/main/kotlin/io/github/kingg22/kobra/builder/finder/BuilderFinder.kt
package io.github.kingg22.kobra.builder.finder

import com.intellij.psi.PsiClass
import org.jetbrains.annotations.VisibleForTesting

public object BuilderFinder {
    @VisibleForTesting
    internal const val SEARCH_PATTERN: String = "Builder"

    public const val EMPTY_STRING: String = ""

    @JvmStatic
    public fun findBuilderForClass(psiClass: PsiClass): PsiClass? {
        val innerBuilderClass: PsiClass? = tryFindInnerBuilder(psiClass)
        if (innerBuilderClass != null) {
            return innerBuilderClass
        } else {
            val className = psiClass.name ?: return null
            return findClass(psiClass, className + SEARCH_PATTERN)
        }
    }

    @JvmStatic
    public fun findClassForBuilder(psiClass: PsiClass): PsiClass? {
        val className = psiClass.name ?: return null
        val searchName = className.replaceFirst(SEARCH_PATTERN.toRegex(), EMPTY_STRING)
        return findClass(psiClass, searchName)
    }

    private fun tryFindInnerBuilder(psiClass: PsiClass): PsiClass? {
        var innerBuilderClass: PsiClass? = null
        val allInnerClasses = psiClass.allInnerClasses
        for (innerClass in allInnerClasses) {
            val innerClassName = innerClass.name ?: continue
            if (innerClassName.contains(SEARCH_PATTERN)) {
                innerBuilderClass = innerClass
                break
            }
        }
        return innerBuilderClass
    }

    private fun findClass(psiClass: PsiClass, searchName: String): PsiClass? = if (typeIsCorrect(psiClass)) {
        ClassFinder.findClass(searchName, psiClass.project)
    } else {
        null
    }

    private fun typeIsCorrect(psiClass: PsiClass): Boolean =
        !psiClass.isAnnotationType && !psiClass.isEnum && !psiClass.isInterface
}
